### PR TITLE
File non string open mode

### PR DESCRIPTION
**Pull request by FreeSlave**
That adds a new way to create and open File object - instead of using C standard library style string modes it allows to use symbolic flags. This is a more granular approach and allows to open file in 'Create New' mode to ensure the file did not exist before opening. Note that C11 standard has 'x' flag that does the same thing, but C11 is not always implemented.

### CODE REVIEW

File: std/stdio.d [hunk 1]
```diff
@@ -466,6 +466,283 @@ Throws: `ErrnoException` if the file could not be opened.
...
+         */
+        existingOnly = 1 << 5,
+    }
+
+    private static bool hasAnyWriteFlag(OpenMode openMode) @safe nothrow pure @nogc
+    {
+        return (openMode & (OpenMode.write | OpenMode.truncate | OpenMode.append | OpenMode.createNew)) != 0;
+    }
+
+    private static const(char)[4] modezFromOpenMode(OpenMode openMode) @trusted
```
Comment (BBasile) [inline std/stdio.d]: this is `@safe`

File: std/stdio.d [hunk 2]
```diff
@@ -466,6 +466,283 @@ Throws: `ErrnoException` if the file could not be opened.
...
+            else if ((openMode & OpenMode.read) != 0 && !hasAnyWriteFlag(openMode))
+                mode |= O_RDONLY;
+
+            auto fd = _open_osfhandle(cast(intptr_t) handle, mode);
+        }
+        errnoEnforce(fd >= 0, "Cannot open Windows HANDLE");
+        return fdToFp(fd, openMode);
+    }
+
+    private static FILE* openFile(string name, OpenMode openMode) @trusted
```
Comment (BBasile) [inline std/stdio.d]: this is `@safe` since inside calls are trusted.

File: std/stdio.d [hunk 3]
```diff
@@ -466,6 +466,283 @@ Throws: `ErrnoException` if the file could not be opened.
...
+    {
+        return (openMode & (OpenMode.write | OpenMode.truncate | OpenMode.append | OpenMode.createNew)) != 0;
+    }
+
+    private static const(char)[4] modezFromOpenMode(OpenMode openMode) @trusted
+    {
+        const bool anyWrite = hasAnyWriteFlag(openMode);
+        char[4] modez;
+        size_t i = 0;
+        if (openMode & OpenMode.read) {
```
Comment (BBasile) [inline std/stdio.d]: curly braces on their own lines

File: std/stdio.d [hunk 4]
```diff
@@ -466,6 +466,283 @@ Throws: `ErrnoException` if the file could not be opened.
...
+            } else {
+                modez[i++] = 'r';
+            }
+        } else if (anyWrite) {
+            if (openMode & OpenMode.append)
+                modez[i++] = 'a';
+            else
+                modez[i++] = 'w';
+        }
+        assert(i < modez.length);
```
Comment (BBasile) [inline std/stdio.d]: boundscheck will catch this too automatically and before.

File: std/stdio.d [hunk 5]
```diff
@@ -466,6 +466,283 @@ Throws: `ErrnoException` if the file could not be opened.
...
+         * Don't create file if it does not exist when opening in write mode.
+         * This flag is not necessary when opening a file in read-only mode.
+         * This flag can't be used together with `OpenMode.createNew` flag.
+         */
+        existingOnly = 1 << 5,
+    }
+
+    private static bool hasAnyWriteFlag(OpenMode openMode) @safe nothrow pure @nogc
+    {
+        return (openMode & (OpenMode.write | OpenMode.truncate | OpenMode.append | OpenMode.createNew)) != 0;
```
Comment (BBasile) [inline std/stdio.d]: how about `with(OpenMode) return...` for a smaller return statement.

File: std/stdio.d [hunk 6]
```diff
@@ -466,6 +466,283 @@ Throws: `ErrnoException` if the file could not be opened.
...
+        else
+            fd = core.sys.posix.fcntl.open(namez, mode);
+        errnoEnforce(fd >= 0);
+        return fd;
+    }
+
+    version(Windows) private static HANDLE openHandle(string name, OpenMode openMode) @trusted
+    {
+        import std.internal.cstring : tempCString;
+        import core.sys.windows.windows;
```
Comment (BBasile) [inline std/stdio.d]: selective import, as you did for the Posix version of the function

File: std/stdio.d [hunk 7]
```diff
@@ -466,6 +466,283 @@ Throws: `ErrnoException` if the file could not be opened.
...
+
+    private static const(char)[4] modezFromOpenMode(OpenMode openMode) @trusted
+    {
+        const bool anyWrite = hasAnyWriteFlag(openMode);
+        char[4] modez;
+        size_t i = 0;
+        if (openMode & OpenMode.read) {
+            if (anyWrite) {
+                if (openMode & OpenMode.append) {
+                    modez[i++] = 'a';
```
Comment (BBasile) [inline std/stdio.d]: This whole block has a poor style. You could return string literals, taking advantage of the automatic null sentinel, eg:

```d
        if (openMode & OpenMode.read)
        {
            if (anyWrite)
            {
                if (openMode & OpenMode.append)
                    return "a+"; // or "a+\0\0" if necessary
                else if (openMode & OpenMode.existingOnly)
                    return "r+";
                else if (openMode & OpenMode.truncate)
                    return "rw";
                else
                   return "w+";
            }
            else return "r";
        }
 // etc....
```

File: std/stdio.d (hunk 6 above)
Comment (FreeSlave) [reply std/stdio.d]: It's too many symbols. But will do if it's really required.

Comment (BBasile) [reply std/stdio.d]: I think you don't have the choice if you want tthis PR to become green Circle CI (still) fails because of this.

**Comment by CyberShadow** [comment]
Hmm, I don't think this will fly (mainly with @andralex).

This adds a lot of new code, while being able to use enums instead of strings is only a superficial improvement.

std.stdio is specifically designed to be built on top of C stdio. Repeating its work by reimplementing things that belong in C stdio is a tough sell as well. I think a more orthogonal approach would be to e.g. implement the x file mode in DigitalMars' libc, which would then make it available on Win32 D.

Is there anything else that I missed?

**Comment by BBasile** [comment]
@CyberShadow  I expected such as reaction, however my review was purely technical. And @FreeSlave, you can still create a DUB package for this, so that work is not lost, in case final decision is "no no no".

**Comment by FreeSlave** [comment]
> Is there anything else that I missed?

I think all useful combinations are covered with string mode considering there's a support for x and I believe it existed in glibc for a long time before C11. But I'm not sure about unix-like platforms that don't use glibc and also Microsoft libraries (fixing dmc C library is not enough, because user may link to Microsoft C library, right?)

**Comment by CyberShadow** [comment]
> But I'm not sure about unix-like platforms that don't use glibc 

Along the lines of the above, it is the burden of said libc libraries to support C11.

> and also Microsoft libraries (fixing dmc C library is not enough, because user may link to Microsoft C library, right?)

They may, yes, if they use `-m64` or `-m32mscoff`.

**Comment by FreeSlave** [comment]
I'm not fond of the fact that the public interface of D standard libary depends on the special features of the C library implementation. But I guess std.stdio is just designed that way. If others say no, I'll take the idea of making the dub package. Fortunately std.stdio has fdopen and windowsHandleOpen publicly available, so it's possible to implement such functionality without changing the phobos code.

**Comment by CyberShadow** [comment]
> Fortunately std.stdio has fdopen and windowsHandleOpen publicly available, so it's possible to implement such functionality without changing the phobos code.

Yes, I added them when I needed to implement something similar to this PR in my own library:
https://github.com/CyberShadow/ae/blob/7ba3063ebff069793dba01fb193ce4d1e667ecd6/sys/file.d#L1022

**Comment by FreeSlave** [comment]
Btw, I made this PR only because you agreed that "Specifying flags via a string is an ancient API anti-pattern" when discussing https://github.com/dlang/phobos/pull/5510

**Comment by schveiguy** [comment]
> We should definitely not use them for std.io should that ever materialize.

FYI: https://github.com/MartinNowak/io uses enums, but also uses strings as a convenience feature to specify the mode (but all is done at compile-time). I like the strings simply because they are short and to the point vs. an enum bitfield where you have to repeat the name constantly.

See: https://martinnowak.github.io/io/std/io/file/mode.html

**Comment by FreeSlave** [comment]
Are we waiting for someone else's decision or I can close this?

**Comment by CyberShadow** [comment]
You can try pinging Andrei/Walter, as they are the ones who have the final say.

**Comment by FreeSlave** [comment]
@andralex can you tell if we need this stuff in phobos?

**Comment by andralex** [comment]
cc @atilaneves

**Comment by atilaneves** [comment]
I think that a dub package would be preferable, so closing.

**Pull request closed without merging**